File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
private:
    long long solve(vector<int> &nums,int k){
        long long sum = 0;
        unordered_map<int,int> ump;
        long long maxSum = sum;
        for(int i=0;i<k;i++){
            sum+=nums[i];
            ump[nums[i]]++;
            if (ump.size() == k)
                maxSum = sum;
        }
        long long currentSum = sum;
        
        for(int i=k;i<nums.size();i++){
            currentSum = currentSum-nums[i-k];
            ump[nums[i-k]]--;
            if(ump[nums[i-k]]==0){
                ump.erase(nums[i-k]);
            }
            currentSum+=nums[i];
            ump[nums[i]]++;
            if(ump.size()==k){
                maxSum = max(maxSum,currentSum);
            }
        }
        return maxSum;
    }
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        long long ans = solve(nums,k);
        return ans;
    }
};

File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        long sum=0;
        Map<Integer,Integer> mp=new HashMap<>();
        for(int i=0;i<k;i++){
            sum+=nums[i];
            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);
        }
        long max=0;
        if(mp.size()==k) max=sum;
       
        for(int i=0;i<nums.length-k;i++){
            mp.put(nums[i],mp.get(nums[i])-1);
            if(mp.get(nums[i])==0){
                mp.remove(nums[i]);
            }
            sum-=nums[i];
            mp.put(nums[k+i],mp.getOrDefault(nums[k+i],0)+1);
            sum+=nums[k+i];
            if(mp.size()==k)
                max=Math.max(max,sum);
        }
        return max;
    }
}